pycbc/vetoes/chisq_cpu.pyx

```diff
@@ -0,0 +1,145 @@
+import numpy
+cimport numpy
+from libc.math cimport cos, sin # This imports c's sin and cos function from the math library
+from cython import wraparound, boundscheck, cdivision
+from pycbc.types import real_same_precision_as
+
+ctypedef fused REALTYPE:
+    float
+    double
+
+ctypedef fused COMPLEXTYPE:
+    float complex
+    double complex
+
+@boundscheck(False)
+@wraparound(False)
+def chisq_accum_bin_cython(numpy.ndarray[REALTYPE, ndim=1] chisq,
+                           numpy.ndarray[COMPLEXTYPE, ndim=1] q, int N):
+    # This can be made parallelized?!
+    for i in range(N):
+        chisq[i] += q[i].real*q[i].real+q[i].imag*q[i].imag
+
+@boundscheck(False)
+@wraparound(False)
+@cdivision(True)
+def point_chisq_code(numpy.ndarray[REALTYPE, ndim=1] chisq,
+                     numpy.ndarray[COMPLEXTYPE, ndim=1] v1,
+                     int n, int slen,
+                     numpy.ndarray[REALTYPE, ndim=1] shifts,
+                     numpy.ndarray[numpy.uint32_t, ndim=1] bins,
+                     int blen):
+    # Do I need to declare UINT vs INT??
+    cdef int num_parallel_regions, bstart, bend, i, j, k, r, start, end
+    cdef numpy.ndarray[REALTYPE, ndim=1] outr, outi, pr, pi, vsr, vsi, outr_tmp, outi_tmp
+    cdef COMPLEXTYPE v
+    cdef REALTYPE vr, vi, t1, t2, k1, k2, k3, vs, va
+
+    #real_type = real_same_precision_as(chisq)
+    real_type = numpy.float64
+    #complex_type = complex_same_precision_as(chisq)
+
+    num_parallel_regions = 16
+
+    outr = numpy.zeros(n, dtype=real_type)
+    outi = numpy.zeros(n, dtype=real_type)
+    pr = numpy.zeros(n, dtype=real_type)
+    pi = numpy.zeros(n, dtype=real_type)
+    vsr = numpy.zeros(n, dtype=real_type)
+    vsi = numpy.zeros(n, dtype=real_type)
+    outr_tmp = numpy.zeros(n, dtype=real_type)
+    outi_tmp = numpy.zeros(n, dtype=real_type)
+
+    for r in range(blen):
+        bstart = bins[r] # int
+        bend = bins[r+1] # int
+        blen = bend - bstart # int
+
+        #outr = numpy.zeros(n, dtype=real_type)
+        #outi = numpy.zeros(n, dtype=real_type)
+        for i in range(n):
+            outr[i] = 0.
+            outi[i] = 0.
+
+
+        # CAN WE MAKE THIS LOOP PARALLEL?!
+        for k in range(num_parallel_regions):
```

Comment:
This can't be made parallel actually if you move the below mallocs up to the top. That's an issue for down the line, but I think the loop over num_parallel_regions doesn't make sense then and can probably be removed. That at least should simplify things.

Reply:
Yeah, I realize that the mallocs would need to move back down again if parallelizing, but wanted to make sure this wasn't affecting performance in the current code.